### ISSUE TITLE
[Bug] Validator to Safety-Rules networking is extremely brittle

### ISSUE THREAD

**Issue by mgorven** [libra/libra#4963]
# 🐛 Bug

We deployed a bugfix to pre-mainnet which involved restarting the validator pod but not the safety-rules pod. 7 nodes got into a state where the validator was stuck communicating with safety-rules. Operators had to manually restart both the validator and safety-rules pods to get them going again.

The validator logs show `Connection established to upstream 10.100.106.73:6185`, but not `RoundManager started`. The safety-rules logs had no messages since the validator restarted.

I believe that safety-rules thought the connection from the previous validator process was still alive, and it only handles one connection at a time. The new validator process established a TCP connection (because this is handled by the kernel on the remote side), but when it tried to communicate with safety-rules there was no response. There are two main problems here:

* safety-rules doesn't handle new connections if it thinks there is an existing one. It should accept new connections, and probably terminate any old ones if we only want it handling one at a time.
* The validator didn't time out waiting for safety-rules. It should have timeouts for all operations, logging and reconnecting.

This communication channel has been the likely cause of numerous issues in the past, we need to take a thorough look at the connection handling and ensure that it is robust.

**Comment by mgorven** [comment]
cc @sherry-x @gregnazario

**Comment by mgorven** [comment]
S204936

**Comment by gregnazario** [comment]
Okay, so this sounds like a combined safety-rules/networking problem of multiple things.
1. Connections are being bottlenecked / blocked by previous connections in safety rules
2. Needs more timeouts

David I know doesn't get back until tomorrow, how urgent / bad was it?

**Comment by sherry-x** [comment]
@gregnazario currently we resolved this issue on Pre-mainnet with a restart for both validator and safety-rule pods, so it's mitigated for now. this can wait until david come back

**Comment by davidiw** [comment]
Thanks for the issue. Couple of notes:
* In most cases, I would have expected peers to cleanly close, any idea what happened to the original validator node? In my local tests, I turned off a validator node and connected a new one and it worked without issue. I would like to be able to repro this.
* In the cases where it doesn't cleanly close, we should expect TCP connections to eventually timeout (https://unix.stackexchange.com/questions/519501/default-tcp-keepalive-settings) however in the circumstances where the issue was being exhibited, the default is unacceptable.

I think moving forward, the easiest solution is to use a 30 second read timeout on an accepted socket. If the read doesn't come in within the expected period of time, tear down the connection and start listening again.

Why not reduce the keepalives? Rust makes it particularly painful to configure those values from within the standard library :/.

I suppose we could probably add a 30 second timeout to all of our trusted interfaces (ureq in vault interface and secure/net) both on read and writes with little issue.

**Comment by mgorven** [comment]
> In most cases, I would have expected peers to cleanly close

In theory this should happen when a process dies. In practice one absolutely cannot rely on this happening. Part of the problem is the lack of graceful shutdown (#4981), but we still must handle processes dying and connections staying open.

> In the cases where it doesn't cleanly close, we should expect TCP connections to eventually timeout 

TCP Keepalives are an optional feature which must be explicitly enabled on connections by both ends. This option doesn't get used very much, generally application level keepalives or timeouts are used instead.

> I think moving forward, the easiest solution is to use a 30 second read timeout on an accepted socket.

That's reasonable because we expect communication between the validator and safety-rules to be continuously active. I think safety-rules still needs to handle new incoming connections immediately though.

**Comment by davidiw** [comment]
Gonna close this out as it has been primarily addressed. The follow up should be handled as we explore future efforts to unify the usage of LibraNet system and, potentially, pull in async into the TCB.

**Comment by mgorven** [comment]
> Gonna close this out as it has been primarily addressed. The follow up should be handled as we explore future efforts to unify the usage of LibraNet system and, potentially, pull in async into the TCB.

Are you confident about all the related error handling and retry/reconnection logic? Can you make another task for safety-rules accepting incoming connections immediately if that isn't done yet.
